doc/website/blog/2021-11-1-sync-to-external-data-systems-using-hooks.md

```diff
@@ -0,0 +1,306 @@
+
+```go title="ent/schema/hooks.go"
+// EnsureImageExists ensures the avatar_url points
+// to a real object in the bucket.
+func EnsureImageExists() ent.Hook {
+	hk := func(next ent.Mutator) ent.Mutator {
+		return hook.UserFunc(func(ctx context.Context, m *ent.UserMutation) (ent.Value, error) {
+			avatarURL, exists := m.AvatarURL()
+			if !exists {
+				return nil, errors.New("avatar field is missing")
+			}
+			switch exists, err := m.Bucket.Exists(ctx, avatarURL); {
+			case err != nil:
+				return nil, fmt.Errorf("check key existence: %w", err)
+			case !exists:
+				return nil, fmt.Errorf("key %q does not exist in the bucket", avatarURL)
+			default:
+				return next.Mutate(ctx, m)
+			}
+		})
+	}
+	return hook.On(hk, ent.OpCreate)
+}
+
+// DeleteOrphans cascades the user deletion to the bucket.
+// Hence, when a user is deleted, its avatar image is deleted
+// as well.
+func DeleteOrphans() ent.Hook {
+	hk := func(next ent.Mutator) ent.Mutator {
+		return hook.UserFunc(func(ctx context.Context, m *ent.UserMutation) (ent.Value, error) {
+			id, exists := m.ID()
+			if !exists {
+				return nil, errors.New("id field is missing")
+			}
+			u, err := m.Client().User.Get(ctx, id)
+			if err != nil {
+				return nil, fmt.Errorf("getting deleted user: %w", err)
+			}
+			if err := m.Bucket.Delete(ctx, u.AvatarURL); err != nil {
+				return nil, fmt.Errorf("deleting user avatar from bucket: %w", err)
+			}
+			return next.Mutate(ctx, m)
+		})
+	}
+	return hook.On(hk, ent.OpDeleteOne)
+}
+```
+
+Now, it's time to test our hooks! Let's write a testable example that verifies that our 2 hooks work as expected.
+
+```go
+package main
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	"github.com/a8m/ent-sync-example/ent"
+	_ "github.com/a8m/ent-sync-example/ent/runtime"
+
+	"entgo.io/ent/dialect"
+	_ "github.com/mattn/go-sqlite3"
+	"gocloud.dev/blob"
+	_ "gocloud.dev/blob/memblob"
+)
+
+func Example_SyncCreate() {
+	ctx := context.Background()
+	// Open an in-memory bucket.
+	bucket, err := blob.OpenBucket(ctx, "mem://photos/")
+	if err != nil {
+		log.Fatal("failed opening bucket:", err)
+	}
+	client, err := ent.Open(
+		dialect.SQLite,
+		"file:ent?mode=memory&cache=shared&_fk=1",
+		// Inject the blob.Bucket on client initialization.
+		ent.Bucket(bucket),
+	)
+	if err != nil {
+		log.Fatal("failed opening connection to sqlite:", err)
+	}
+	defer client.Close()
+	if err := client.Schema.Create(ctx); err != nil {
+		log.Fatal("failed creating schema resources:", err)
+	}
+	if err := client.User.Create().SetName("a8m").SetAvatarURL("a8m.png").Exec(ctx); err == nil {
+		log.Fatal("expect user creation to fail because the image does not exist in the bucket")
+	}
+	if err := bucket.WriteAll(ctx, "a8m.png", []byte{255, 255, 255}, nil); err != nil {
+		log.Fatalf("failed uploading image to the bucket: %v", err)
+	}
+	fmt.Printf("%q\n", keys(ctx, bucket))
```

Comment:
It's easy enough to understand what `keys` does, but I wonder if you left it out intentionally

Reply:
Yes